src/server/frontend_wayland/wl_surface.h

```diff
@@ -183,6 +187,7 @@ class WlSurface : public wayland::Surface
     std::optional<std::vector<mir::geometry::Rectangle>> input_shape;
     std::vector<SceneSurfaceCreatedCallback> scene_surface_created_callbacks;
     wayland::Weak<Viewport> viewport;
+    FractionalScaleV1* fractional_scale{nullptr};
```

Comment:
Hm. This should be a `wayland::Weak<FractionalScaleV1>` - the lifetime of Wayland objects is outside of our control, and the client might destroy the `fractional_scale` without destroying the surface, which would leave this a dangling pointer.